### ISSUE TITLE
Stage lines

### ISSUE THREAD

**Issue by kissgyorgy** [jesseduffield/lazygit#11]
I have been waiting for something like this for a long time, the only feature I'm missing is the ability to select lines and only stage them without staging the whole file.

**Comment by jesseduffield** [comment]
Yeah that's something I very much want to add. I tried doing an interactive shell session with `git add -p` but couldn't quite get it working. It's definitely something I want in the app

**Comment by v3rm0n** [comment]
I would definitely try to switch from [tig](https://jonas.github.io/tig/) if I could stage lines, lazygit looks much nicer.

**Comment by from-nibly** [comment]
+1 I think this would get me to switch from git kraken.

**Comment by jesseduffield** [comment]
I've just pushed a commit to master which allows you to hit 't' and get redirected to a subprocess of `git add --patch` for the selected file. It's not as good as a UI in the app itself, but it's a start :)

**Comment by kissgyorgy** [comment]
> but it's a start :)

No it's not. It's nowhere near the consistency and quality to the current version at all. There is no step from a modest solution like that to a pleasant UI like lazygit is currently.
Please don't poison a great project with shitty solutions like that.

**Comment by fiendish** [comment]
@kissgyorgy hey, now. It's no worse than the current alternative, which is to switch to a different program for this functionality. This is just a shortcut for that. I'm sure that @jesseduffield also wants to find a way to integrate this into the UI, but I for one would rather not have nothing in the meantime.

**Comment by jesseduffield** [comment]
@kissgyorgy we differ in our ideologies here. As I originally made lazygit to help me with git, the mentality has always been to iteratively improve without breaking any existing features. Switching to the `git add -p` subprocess is not as good as having a subprocess communicated to via the UI, as I stated in my above comment, but I have already made use of it in my own use of the app, and the switching-to-subprocess pattern will be around for various other things like committing when a password is required, or switching to vi, for some time until we work out how to integrate properly with the subprocesses. Though I want lazygit to be a tool with a highly polished UI, it's still in early days (not v1.0.0 yet) and so tradeoffs like this will happen as we iron out the kinks.

Having said that, if you ever post a comment with a tone like that again (removed swear word notwithstanding) I'm blocking you from the repo. It's not a nice thing to wake up to.

**Comment by kissgyorgy** [comment]
Yea, sorry about the tone, it was not very nice, and I removed it immediately when I realized.
But as I told earlier, I was searching for a long time for something like this and would be sad to see quality diminish after a great start.

I didn't looked into it before, but I see you are running git commands with os/exec, but wouldn't it be easier using a libgit2 wrapper? It's also more efficient for sure because you don't have the subprocess and text parsing overhead.

**Comment by kissgyorgy** [comment]
> but wouldn't it be easier using a libgit2 wrapper

Ok, I see that go-git is a pure-Go Git manipulator library, but I'm just curious then, why do you need os/exec calls? Are there missing commands from go-git?

**Comment by jesseduffield** [comment]
All good, and yeah I see your perspective as well. Regarding go-git, pretty much all the calls to git right now can be swapped out for go-git (I only became aware of that project recently). I've gotten started on that process, e.g. in this commit https://github.com/jesseduffield/lazygit/commit/4832d365f1a2067c3bc464ec07b19f343ee4c495
But some things will take a little more time because the code depends on the format of the output. Looking at the go-git examples, it seems like there's some good stuff for the async operations that we could use. Having said that, their support for git add is limited, so we'd need to come up with a solution that involves making a patch file based on user input and then applying that.

**Comment by kissgyorgy** [comment]
> process, e.g. in this commit 4832d36

Very nice!

**Comment by jesseduffield** [comment]
I've got a [PR](https://github.com/jesseduffield/lazygit/pull/338) up for this feature now. Please feel free to test the PR, I'll be merging it soon
